supplier_product/models.py

```diff
@@ -0,0 +1,67 @@
+from django.db import models
+from django.core.validators import MinLengthValidator, MinValueValidator
+from product.models import Product
+
+
+class User(models.Model):
+    user_name = models.CharField(max_length=32,
+                                 primary_key=True,
+                                 null=False,
+                                 validators=[MinLengthValidator(6)],
+                                 help_text="User Name")
+    first_name = models.CharField(max_length=32,
+                                  validators=[MinLengthValidator(1)],
+                                  help_text="First Name")
+    last_name = models.CharField(max_length=32,
+                                 validators=[MinLengthValidator(1)],
+                                 help_text="Last Name")
+    password = models.CharField(max_length=16,
+                                validators=[MinLengthValidator(6)],
+                                null=False,
+                                help_text="Password")
+    user_type = models.CharField(max_length=16,
+                                 validators=[MinLengthValidator(6)],
+                                 null=False,
+                                 choices=[('client', 'client'), ('supplier', 'supplier')],
+                                 help_text="User Type - supplier/client")
+
+    class Meta:
+        abstract = True
+
+
+class Supplier(User):
+    """Supplier class inherit from User abstract class.
+    Supplier class lets the app user to do variety of actions
+    which evolves the Product and Delivery class.
+    Attributes:
+        business_name: a string which holds the Supplier business name.
+    """
+    business_name = models.CharField(max_length=32,
+                                     validators=[MinLengthValidator(6)],
+                                     null=False,
+                                     help_text="Business Name")
+
+
+class SupplierProduct(models.Model):
+    suppler_product_id = models.IntegerField(primary_key=True)
+    qr_code = models.ForeignKey(Product, on_delete=models.CASCADE)
+    user_name = models.ForeignKey('Supplier', on_delete=models.CASCADE)
+    price = models.IntegerField(validators=[MinValueValidator(1)])
+    quantity = models.IntegerField(validators=[MinValueValidator(1)])
```

Comment:
couldnt it be 0?

Reply:
Why do we need to hold a complete row for something that is not related to the supplier

Reply:
What do you mean @thetaxas2468? 
The quantity field specifies how many available items (of this specific product) the supplier offers to sell.

Reply:
Lets suppose specific supplier had 1k items for sale and all of them has been sold, now we really need extra 1k rows in our DB with items that was sold by the supplier and now their quantity are 0?

Reply:
Oh, you mean to delete the products with a quantity equal to 0. 
First, the DB knows how to handle big amounts of data. That's why we use it. 
You are suggesting deleting rows from the DB when they are no longer in use. 
But what about saving the history of each supplier? 
Wouldn't a supplier be able to see the products he sold in the past? 
Usually what we do is save the records with a field to specify it is no longer in use (like quantity=0, status=0, available = false, etc.).

Reply:
> Oh, you mean to delete the products with a quantity equal to 0. First, the DB knows how to handle big amounts of data. That's why we use it. You are suggesting deleting rows from the DB when they are no longer in use. But what about saving the history of each supplier? Wouldn't a supplier be able to see the products he sold in the past? Usually what we do is save the records with a field to specify it is no longer in use (like quantity=0, status=0, available=false, etc.).

we havent really discussed about the option to look into the history of a product nor a supplier... due to the schedule i think we can all agree not to add new features (see the history of supplier/product), so deleting the row will be the best option atm.
would like to hear if you agree or not @nofaralfasi

Reply:
I agree with you about not adding the feature of viewing the seller's history. We shouldn't add more features to our projects.
However, if we delete a `supplier_product` object from the DB when its quantity is zero, we would have a problem with the `orderd_product`.
If we delete a `supplier_product` that is assigned to an `orderd_product` object, we would either have a null FK (for the `supplier_product_id`), or the `orderd_product` object will be deleted (depending on our `on_delete` value).

Reply:
i think its a matter of good design, if we think ahead, ordered product should have cascade on delete (as we did).
i assume that youre saying that we should make sure the ordered product wont be null?

Reply:
No, the opposite. 
The `ordered_product` has a foreign key `supllier_product_id`. 
If we delete the `supllier_product`, we would get a null value for the `supllier_product_id`  (if we use `on_delete = SET_NULL`), or the `ordered_product` object will be deleted (if we use `on_delete = CASCADE`).
Both options are not what we want, therefore I think we shouldn't delete the `supllier_product`.

Reply:
but if the `supplier_product` is deleted, it means the product the supplier is selling is no longer available, which means we need to delete it from the `ordered_product`, therefor `on_delete = CASCADE`, is the right option, no?

Reply:
I see your point. 
You are making the assumption that the `supplier_product` quantity is set to zero after the order is delivered. 
I don't think it is a good practice but if the other mentors agree to it, then so be it.